src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -410,27 +409,20 @@ class SchedulerActions(
     instancesToKill.foreach { instances: Seq[Instance] =>
       logger.info(s"Scaling ${runSpec.id} from ${runningInstances.size} down to $targetCount instances")
 
-      launchQueue.purge(runSpec.id)
-        .recover {
-          case NonFatal(e) =>
-            logger.warn("Async purge failed: {}", e)
-            Done
-        }.foreach { _ =>
-          logger.info(s"Killing instances ${instances.map(_.instanceId)}")
-
-          def killInstances(instances: Seq[Instance]): Future[Done] = async {
-            val changeGoalsFuture = instances.map { i =>
-              if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped)
-              else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned)
-            }
-
-            await(Future.sequence(changeGoalsFuture))
-            await(killService.killInstances(instances, KillReason.OverCapacity))
-          }
+      async {
+        await(launchQueue.purge(runSpec.id))
 
-          killInstances(instances)
+        logger.info(s"Adjusting goals for instances ${instances.map(_.instanceId)} (${GoalChangeReason.OverCapacity})")
+        val instancesAreTerminal = killService.watchForKilledInstances(instances)(mat)
+        val changeGoalsFuture = instances.map { i =>
+          if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped, GoalChangeReason.OverCapacity)
+          else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned, GoalChangeReason.OverCapacity)
         }
+        await(Future.sequence(changeGoalsFuture))
+        await(instancesAreTerminal)
```

Comment:
I know, that this was the semantic previously, but why do we need to wait here for the instances to be killed? What's the reason to not simply `killInstancesAndForget`?